Fixed every lags of the servers list

boubouleuh opened the pull request:
I was so bored of the lags so i decided to fix them all. 

Everything i added was strongly tested. I got some help from AI to understand angularjs stuff and that helped me to gain times. 

I confirm that i didnt copy pasted some bad codes coming from AI or dumb peoples on the internet.

That PR features many performances improvement with ng-repeat track by and the infinite scroll using angular-inview. I also removed every unoptimized codes

boubouleuh committed:
test

boubouleuh committed:
Performance improve by using ng-repeat

boubouleuh committed:
fixed ng-repeat, added infinite scroll

boubouleuh commented:
Theres just an issue with the scrollbar being weird for some reasons, but i cant understand why

boubouleuh committed:
fixed error on background, fixed scrollbar, fixed container height, fixed favorites being undefined

boubouleuh committed:
fixed background color behavior

boubouleuh commented:
That was the last fix I needed to do so it behavior the same as before.

So for me it seems ready I don't have more work to do unless theres some suggestions or changes needed after review.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -784,7 +804,7 @@ function($scope, $state, $timeout, $filter) {
 
 	// Resize the server list
-	setServersTableHeight();
+	// setServersTableHeight();
```
Why was this commented out? If it's no longer needed you can just remove it.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -748,14 +748,34 @@ function($scope, $state, $timeout, $mdDialog, $filter, ConfirmationDialog, toast
 .controller('MultiplayerServersController', ['$scope', '$state', '$timeout', '$filter',
 function($scope, $state, $timeout, $filter) {
 
+	// document.getElementById('serverList').parentElement.style.overflow = "hidden";
```
Why was this commented out? If it's no longer needed you can just remove it.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -1217,6 +1279,18 @@ function returnDefault(data, type) {
 	else return data;
 }
 
+function infinitScrollController($scope) {
+
+  $scope.limit = 10;
+
+  $scope.loadMore = function (last, inview) {
+		if (last && inview) {
+			$scope.limit += 10;
+		}
+	
+  }
+}
```
Could the indenting be corrected here?

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -1318,11 +1393,12 @@ globalThis.openExternalLink = function(url){
 	bngApi.engineLua(`MPCoreNetwork.openURL("`+url+`")`);
 }
 
-function getServerInfoHTML(d) {
+function getServerInfoHTML(c) {
```
Why was this changed?

boubouleuh replied:
I now renamed it "row" since it doesnt represent the original data object anymore, so i get it after with the "d" variable.
'c' was a bad name i fixed it

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -1437,11 +1490,11 @@ async function populateTable(tableTbody, servers, tab, searchText = '', checkIsE
-		// Set the color relative to either favorite, featured, official or normal
-		var bgcolor = isFavorite && type == 0 ? 'rgba(255, 215, 0, 0.35)!important' : server.featured ? 'rgba(0, 128, 0, 0.25)!important' : server.official ? 'rgba(255, 106, 0, 0.25)!important' : server.partner ? 'rgba(0, 123, 195, 0.3)!important' : 'rgba(0, 0, 0, 0)!important';
 
+		// $scope.serversTable.push(["server"=server, "bgcolor"=bgcolor, "isFavorite"=isFavorite, "isRecent"=isRecent, "sname"=server.sname]);
```
See previous comment about the comments, also for line 1497,1520, 1526 and 1621

WiserTixx commented on ui/modModules/multiplayer/angular-inview.js:
```diff
@@ -0,0 +1,389 @@
+// # Angular-Inview
```
Could this file be put into the `ui/lib/ext/` folder and could a header be added with the license information?

WiserTixx reviewed (changes requested):
My initial review, haven't properly tested this PR yet.

boubouleuh committed:
removing useless comments, renaming getServerInfoHTML parameter, moved anguar-inview.js and added licence header

boubouleuh commented:
The infinite scroll limit can be changed if needed;I also found a possible issue while scrolling down, if the name of a server is longer than another it can weirdly change the size of every other server name box (It probably does the same for the map name and the player count.) , resulting in a possible unwanted behavior so the fix would be to set the size of the name box and others, but if the server name is longer than expected then it will probably cut it, is that an issue?

boubouleuh committed:
fixed td width on server items, increased limit steps, removing useless overflow-y

boubouleuh commented:
I cant do more to the scrollbar teleporting a little, it doesn't affect the scroll itself too much I think? ; it exists something like that: ``in-view-options="{ viewportOffset: ['-50%', 0, 0]}"`` that can maybe be smoother but I can't manage to make it work sadly.

boubouleuh committed:
Fixed fav button not refreshing when adding or removing the fav from the ServerInfoRow, ServerInfoRow are now managed by Angularjs, fixed updating background color on fav add / remove, selected row are automatically unselected on view change

boubouleuh commented:
I fixed some randoms issue that was standing for a long time because why not.

It will probably need another review :v

WiserTixx commented:
> I cant do more to the scrollbar teleporting a little, it doesn't affect the scroll itself too much I think? ; it exists something like that: `in-view-options="{ viewportOffset: ['-50%', 0, 0]}"` that can maybe be smoother but I can't manage to make it work sadly.

What if you hide the existing scrollbar and then implement a custom one?

boubouleuh commented:
> > I cant do more to the scrollbar teleporting a little, it doesn't affect the scroll itself too much I think? ; it exists something like that: `in-view-options="{ viewportOffset: ['-50%', 0, 0]}"` that can maybe be smoother but I can't manage to make it work sadly.
> 
> What if you hide the existing scrollbar and then implement a custom one?

I can maybe try to do simpler by calculating the final height of the scroll box so the scroll box doesn't change the height at all while scrolling and that would fix that, but idk if it works well sooo i will try some stuff tomorow i hope i can find a way

boubouleuh committed:
not using angular-inview anymore, fixing scrollbar UX and even more improved scrolling performance

boubouleuh commented:
I fixed the issue by doing the infinite scroll manually with onscroll instead of in-view so I removed angular-inview and it works like a charm and its even more better because when you scroll back up it unload the one you don't see anymore. Im kinda proud of this one its easily reusable on others projects.

boubouleuh committed:
fixed height of the list

boubouleuh committed:
fixed undefined value when scrolltop reset on view change

boubouleuh commented:
I finished the last fix .

Let me know what you think.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -794,6 +798,106 @@ function($scope, $state, $timeout, $filter) {
+		let server = finalserver.server;
+		let row = event.target.closest('tr');
+		var table = document.getElementById("serversTable");
+		// Deselect the old row
+		deselect(table.selectedRow);
+
+		// Select the new one
+
+		row.classList.add("highlight");
+		row.selected = true;
+		table.selectedRow = row;
+		
+		// Add the highlight menu
+		highlightedServer = server; // Set it as the selected server
+		//Create and insert the server info tr
```
Could the comments here be removed?

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -794,6 +798,106 @@ function($scope, $state, $timeout, $filter) {
+		window.requestAnimationFrame(() => {
+            $scope.limit = newLimit;
+            
+            // 4. Calcul direct sans Math.max (plus rapide)
+            const remainingHeight = ($scope.serversArray.length - newLimit) * rowHeight;
+            document.getElementById('TEMPSERVERITEM').style.height = 
+                (remainingHeight > 0 ? remainingHeight : 0) + "px";
+            
+            // 5. Forçage ciblé de la mise à jour
```
Could these comments be translated or removed?

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -794,6 +798,106 @@ function($scope, $state, $timeout, $filter) {
+		if (newLimit > $scope.limit) {
+			$scope.limit = newLimit;
+		}
+		window.requestAnimationFrame(() => {
+            $scope.limit = newLimit;
```
Could the intentation be fixed?

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -794,6 +798,106 @@ function($scope, $state, $timeout, $filter) {
+	};
+
+
+	$scope.limit = 48;
+	$scope.loadMore = function () {
+		const rowHeight = 24;
+		const container = document.getElementById('serverListMainContainer').parentElement;
+		if (!container) return;
+
+		const scrolledRows = Math.floor(container.scrollTop / rowHeight);
+
+		const newLimit = scrolledRows + 50;
+		if (newLimit > $scope.limit) {
+			$scope.limit = newLimit;
+		}
```
Could the limit be calculated based on screensize. That way it can be insured that it will always show the proper amount of rows on big screens and not too many on small screens.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -851,13 +956,16 @@ function($scope, $state, $timeout, $filter) {
 	});
 
 	vm.repopulate = async function () {
+		console.log("Repopulating server list")
```
Could this log message be removed?

WiserTixx reviewed (changes requested):
This PR greatly improves the intial load time, but it seems to be slower than normal when scrolling all the way to the bottom. I believe this is partially due to CEF, but maybe it can be eased by also unload the rows above which aren't visible, if that is doable.

Also, currently the search bar and filters are pinned to the top, but this seems to have been broken by this PR. Could you have a look at that?

boubouleuh committed:
calculate infinite scroll limit by using the screen height, removed useless stuff

boubouleuh committed:
fixing indentation

boubouleuh commented:
> This PR greatly improves the intial load time, but it seems to be slower than normal when scrolling all the way to the bottom. I believe this is partially due to CEF, but maybe it can be eased by also unload the rows above which aren't visible, if that is doable.
> 
> Also, currently the search bar and filters are pinned to the top, but this seems to have been broken by this PR. Could you have a look at that?

im trying to fix theses issues right now

WiserTixx commented on ui/modModules/multiplayer/multiplayer.js:
```diff
@@ -1344,9 +1432,14 @@ function getServerInfoHTML(d) {
-						<md-button id="serverconnect-button" class="button servers-button md-button md-default-theme" ng-class="" ng-click="multiplayer.connect(` + d.ip + `, ` + d.port + `)" style="margin-left: 10px;">Connect</md-button>
-						` + favButton + `
+					<div class="row" style="padding-left: 10px;" ng-switch="selected_row.isFavorite">
+						<md-button id="serverconnect-button" class="button servers-button md-button md-default-theme" ng-class="" ng-click="connect(selected_row.server.ip, selected_row.server.port, selected_row.server.strippedName, selected_row.server.isOfficial)" style="margin-left: 10px;">Connect</md-button>
```
```suggestion
						<md-button id="serverconnect-button" class="button servers-button md-button md-default-theme" ng-class="" ng-click="connect(selected_row.server.ip, selected_row.server.port, selected_row.server.strippedName, selected_row.server.official)" style="margin-left: 10px;">Connect</md-button>
```

WiserTixx reviewed (changes requested):
Small change that fixes an issue related to the mod warning for official servers

boubouleuh committed:
fixed official server row that was causing issues

Co-authored-by: Tixx <83774803+WiserTixx@users.noreply.github.com>

boubouleuh committed:
fixed the pinned serversTableHead ; increasing performances again by only rendering the visible rows

boubouleuh committed:
Merge branch 'development' of https://github.com/boubouleuh/BeamMP into development

boubouleuh commented:
hopefully it's fixed, theres still a little of stutters when scrolling very fast with the scrollbar but thats angularjs fault 😭 

So let me know if anything is wrong

WiserTixx commented:
The development team reviewed this PR once again and this is the list of issues that we found.

- The player count can sometimes disappear when scrolling with the ui resolution set to 720p
- Loading animation doesn't cover the whole screen on large resolutions (This was already an issue before but this has partially been changed in this PR)
- When scrolling with a server selected, if the details menu goes partially out of view the serverlist jumps down and the menu disappears (I suppose this can be fixed by taking the details menu into account when checking if a server should be removed)

Besides that everything looks good and this PR greatly improves performance.

boubouleuh committed:
fixed column overflowing on smaller screen size, fixed loading animation on bigger screen size

boubouleuh commented:
I didnt fixed that yet :

> When scrolling with a server selected, if the details menu goes partially out of view the serverlist jumps down and the menu disappears (I suppose this can be fixed by taking the details menu into account when checking if a server should be removed)

I tried something that I think look better, the usage of border-collapse that remove the lines in between the cells, you can compare with and without the ``border-collapse`` but I think its a little better with it. So let me know if i need to change that back.

```css
table#serversTable {
  border-collapse: collapse;
  //other styles
}
```

For the scrolling i have some issues since i need to recalculate some stuff just because of the infoRow when its selected. That will maybe take some times since im not that good at math :(

boubouleuh committed:
Fixed server being unselected while scrolling, fixed scroll jump

boubouleuh commented:
![image](https://github.com/user-attachments/assets/e0732562-7491-4f76-aad0-0a55d9f46b5a)

I did a little design to show how that can look. I'm not a designer, so it's not perfect, but the idea is here.

boubouleuh committed:
fully fixed the scrolling issues by rewriting the entire thing

boubouleuh commented:
Forgot about my last message, I fixed all the stuff with the current way and tested it for an hour, so if I forgot something, let me know! The new code can be hard to understand, but it works, so I assume that's sufficient :v 

Can't wait for this to be merged; JavaScript is haunting me 😭

boubouleuh committed:
removing useless stuff

boubouleuh committed:
fixed recents being incorrect

boubouleuh committed:
Logo Padding Inconsistency in Server List fix

boubouleuh committed:
small performance improve, fixed sortTable by sorting in a different way

boubouleuh commented:
> I have noticed that the sorting seems to be broken. When attempting to sort by players it gives me this error in the UI console.
> 
> ```
> multiplayer.js:1576 Uncaught TypeError: Cannot read properties of undefined (reading 'players')
>     at multiplayer.js:1576:29
>     at Array.sort (<anonymous>)
>     at globalThis.sortTable (multiplayer.js:1575:10)
>     at HTMLTableCellElement.onclick (index.html:1:1)
> ```
> 
> Could this be resolved?

I wasn't able to reproduce that error, sadly, but I still modified the way the servers are sorted by directly modifying the array. So, tell me if it works well. It seems good to me, though.

WiserTixx commented on ui/modModules/multiplayer/multiplayer.css:
```diff
@@ -420,3 +423,68 @@ md-slider.ng-pristine.ng-untouched.ng-valid.md-default-theme.ng-empty {
 }
+
+
+.server-maxplayers {
+  width: 8%; 
+}
+.server-mapname {
+  width: 20%; 
+}
+.server-servername {
+  width: 64%; 
+}
+.server-flag {
+  width: 8%; 
+}
```
```suggestion
.server-maxplayers {
  width: 90px; 
}
.server-mapname {
  width: 20%; 
}
.server-servername {
  width: 70%; 
}
.server-flag {
  width: 90px; 
}
```

WiserTixx commented:
Me and another developer are still unable to sort the serverlist by location, title map and players by clicking on them in the table header.
Error for reference:
```
multiplayer.js:1576 Uncaught ReferenceError: $scope is not defined
    at multiplayer.js:1576:2
    at HTMLTableCellElement.onclick (VM2690 menu.multiplayer.servers:1:1)
```

boubouleuh committed:
Fixed sort, css adjustments

WiserTixx commented:
> There's just one thing: the list sorts only if the selected filters result in a non-empty server list. Is that an issue?

I suppose that's fine, as it would be sorting nothing anyway.
I have noticed that if you select filters that would equate to an empty serverlist it doesn't update the view. Could this be addressed so that it shows nothing if the selected filters equate to nothing?

boubouleuh committed:
filter fix

boubouleuh commented:
> > There's just one thing: the list sorts only if the selected filters result in a non-empty server list. Is that an issue?
> 
> I suppose that's fine, as it would be sorting nothing anyway. I have noticed that if you select filters that would equate to an empty serverlist it doesn't update the view. Could this be addressed so that it shows nothing if the selected filters equate to nothing?
> 
> Also this PR is very close to being merged, and the dev team is very happy with how it's turned out.

I fixed the filter issues.

I'm also happy; That's my first real PR, and it's becoming much better than I expected in the beginning. And probably not the last one on this repository, so I'm happy to help even if I'm still a beginner :3

boubouleuh committed:
fixed beammpchat applying scrollbar style on all others apps including the serverlist

boubouleuh committed:
fixed the playerlist having a visible scrollbar by adding the webkit-scrollbar style instead of being applied by the beammpchat

WiserTixx commented on ui/modModules/multiplayer/multiplayer.css:
```diff
@@ -420,3 +423,67 @@ md-slider.ng-pristine.ng-untouched.ng-valid.md-default-theme.ng-empty {
 }
+
+
+.server-maxplayers {
+  width: 90px; 
+}
+.server-mapname {
+  width: 20%; 
+}
+.server-servername {
+  width: 70%; 
+}
+.server-flag {
+  width: 90px; 
+}
```
```suggestion
.server-maxplayers {
  width: 8%; 
}
.server-mapname {
  width: 20%; 
}
.server-servername {
  width: 64%; 
}
.server-flag {
  width: 8%; 
}
```
Final review, this one is my bad. The change I'm reverting made it initially look better but ended up causing issues with the details window. With percentages it can look a little funny on wide monitors but this is something that doesn't seem very simple to fix.

boubouleuh replied:
I can do different styles depending on the resolution with some media queries for each resolution. It takes time, and I don't really know how it needs to look at very wide resolutions, so if you have a little time to send the style needed for each resolution that I need to fix, I would be very grateful! Otherwise, I can just commit this, but I feel like it's probably too imperfect since a lot of people have a widescreen to play.

In theory, I can also change the rows height depending on the resolution. (That mean being able to change the font size etc..)

WiserTixx replied:
Alright, the flag and player count should both be 90px, the server name should take up 2/3 of the remaining space and the map 1/3.

boubouleuh committed:
Fixed columns width for every resolutions

WiserTixx commented:
Thank you for your contribution, I'm looking forward to seeing this in the next BeamMP update.
